core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/sync/MixedPackageVersionService.java

```diff
@@ -0,0 +1,249 @@
+                .orElse(false);
+    }
+
+    private boolean componentVersionsAreMatchesWithImage(Image image, Set<ClouderaManagerProduct> productsFromImage, String currentCmVersion,
+            Set<ParcelInfo> activeParcels) {
+        return currentCmVersion.equals(getCmVersion(image.getPackageVersions())) && allActiveParcelVersionMatches(activeParcels, productsFromImage);
+    }
+
+    private boolean allActiveParcelVersionMatches(Set<ParcelInfo> activeParcels, Set<ClouderaManagerProduct> productsFromImage) {
+        return activeParcels.stream()
+                .allMatch(parcelInfo -> productsFromImage.stream()
+                        .anyMatch(product -> product.getName().equals(parcelInfo.getName()) && product.getVersion().equals(parcelInfo.getVersion())));
+    }
+
+    private Set<ClouderaManagerProduct> getProductsFromImage(Image image) {
+        return clouderaManagerProductTransformer.transform(image, true, true);
+    }
+
+    private Optional<StatedImage> findCurrentImage(Long stackId) {
+        try {
+            com.sequenceiq.cloudbreak.cloud.model.Image image = imageService.getImage(stackId);
+            return Optional.of(imageCatalogService.getImage(image.getImageCatalogUrl(), image.getImageCatalogName(), image.getImageId()));
+        } catch (CloudbreakImageNotFoundException | CloudbreakImageCatalogException e) {
+            LOGGER.warn("Current image not found for this cluster.");
+            return Optional.empty();
+        }
+    }
+
+    private String createSuggestedVersionsMessage(Image image, Set<ParcelInfo> activeParcels) {
+        String suggestedParcelsMessage = getProductsFromImage(image).stream()
+                .filter(product -> activeParcels.stream().anyMatch(parcelInfo -> parcelInfo.getName().equals(product.getName())))
+                .map(product -> String.format("%s %s", product.getName(), product.getVersion()))
+                .collect(Collectors.joining(", "));
+        return String.format("Cloudera Manager %s, %s", getCmVersion(image.getPackageVersions()), suggestedParcelsMessage);
+    }
+
+    private Set<Image> getImageByCurrentCmVersion(String currentCmVersion, Set<Image> candidateImages) {
+        return candidateImages.stream()
+                .filter(image -> currentCmVersion.equals(getCmVersion(image.getPackageVersions())))
+                .collect(Collectors.toSet());
+    }
+
+    private boolean allActiveParcelVersionMatchesWithCandidateImage(Set<ParcelInfo> activeParcels, String cmVersion, Set<Image> images) {
+        return images.stream().anyMatch(image -> cmVersion.equals(getCmVersion(image.getPackageVersions())) && matchParcelVersions(activeParcels, image));
+    }
+
+    private boolean matchParcelVersions(Set<ParcelInfo> activeParcels, Image image) {
+        Set<ClouderaManagerProduct> clouderaManagerProducts = getProductsFromImage(image);
+        return activeParcels.stream()
+                .allMatch(parcelInfo -> clouderaManagerProducts.stream()
```

Comment:
again, a `HashMap` of name and version would be better here instead of `Set<clouderaManagerProducts>`